src/errors.rs

```diff
@@ -1,74 +1,212 @@
+use crate::{i18n::*, lemmy_errors::LemmyErrorType};
+use core::num::ParseIntError;
+use leptos::*;
 use serde::{Deserialize, Serialize};
 use serde_urlencoded::ser;
-use std::num::ParseIntError;
-use thiserror::Error;
+use strum_macros::{Display, EnumIter};
 
-#[derive(Debug, Clone, Error, Serialize, Deserialize)]
-pub enum LemmyAppError {
-  #[error("Not Found")]
+pub type LemmyAppResult<T> = Result<T, LemmyAppError>;
+
+#[derive(Default, Display, Debug, Clone, Serialize, Deserialize, PartialEq, EnumIter)]
+#[serde(tag = "error", content = "message", rename_all = "snake_case")]
+pub enum LemmyAppErrorType {
+  #[default]
+  Unknown,
   NotFound,
-  #[error("Internal Server Error")]
   InternalServerError,
-  #[error("Couldn't parse params")]
   ParamsError,
-  #[error("{error:?}")]
-  APIError { error: String },
+
+  ApiError(LemmyErrorType),
+
+  EmptyUsername,
+  EmptyPassword,
+  MissingToken,
+
+  MissingReason,
+}
+
+pub fn message_from_error(error: &LemmyAppError) -> String {
+  let i18n = use_i18n();
+
+  let s = match error {
+    LemmyAppError {
+      error_type: LemmyAppErrorType::ApiError(LemmyErrorType::IncorrectLogin),
+      ..
+    } => t!(i18n, invalid_login)().to_string(),
+    LemmyAppError {
+      error_type: LemmyAppErrorType::EmptyUsername,
+      ..
+    } => t!(i18n, empty_username)().to_string(),
+    LemmyAppError {
+      error_type: LemmyAppErrorType::EmptyPassword,
+      ..
+    } => t!(i18n, empty_password)().to_string(),
+    LemmyAppError {
+      error_type: LemmyAppErrorType::MissingReason,
+      ..
+    } => t!(i18n, empty_reason)().to_string(),
+    LemmyAppError {
+      error_type: LemmyAppErrorType::InternalServerError,
+      ..
+    } => t!(i18n, internal)().to_string(),
+    LemmyAppError {
+      error_type: LemmyAppErrorType::Unknown,
+      ..
+    } => t!(i18n, unknown)().to_string(),
+    _ => "t!(i18n, unknown)()".to_string(),
+  };
+
+  logging::log!("{s}");
+
+  s
 }
 
-impl LemmyAppError {
-  // pub fn status_code(&self) -> StatusCode {
-  //   match self {
-  //     LemmyAppError::NotFound => StatusCode::NOT_FOUND,
-  //     LemmyAppError::InternalServerError => StatusCode::INTERNAL_SERVER_ERROR,
-  //     Lemmy
-  //   }
-  // }
+#[derive(Clone, Serialize, Deserialize)]
+pub struct LemmyAppError {
+  pub error_type: LemmyAppErrorType,
+  pub content: String,
 }
 
-pub type LemmyAppResult<T> = Result<T, LemmyAppError>;
+impl serde::ser::StdError for LemmyAppError {}
+
+impl core::fmt::Debug for LemmyAppError {
+  fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+    f.debug_struct("debug LemmyAppError")
+      .field("error_type", &self.error_type)
+      .field("content", &self.content)
+      .finish()
+  }
+}
```

Comment:
How is this different from what you'd get from deriving `Debug`?

Reply:
this approach was adopted from Lemmy server, just gives you options when debugging in future incorporating traces and other information